pkg/controller/security_group.go

```diff
@@ -0,0 +1,443 @@
+		return nil
+	}(obj)
+
+	if err != nil {
+		utilruntime.HandleError(err)
+		return true
+	}
+	return true
+}
+
+func (c *Controller) processNextAddOrUpdateSgWorkItem() bool {
+	obj, shutdown := c.addOrUpdateSgQueue.Get()
+
+	if shutdown {
+		return false
+	}
+
+	err := func(obj interface{}) error {
+		defer c.addOrUpdateSgQueue.Done(obj)
+		var key string
+		var ok bool
+		if key, ok = obj.(string); !ok {
+			c.addOrUpdateSgQueue.Forget(obj)
+			utilruntime.HandleError(fmt.Errorf("expected string in workqueue but got %#v", obj))
+			return nil
+		}
+		if err := c.handleAddOrUpdateSg(key); err != nil {
+			c.addOrUpdateSgQueue.AddRateLimited(key)
+			return fmt.Errorf("error syncing '%s': %s, requeuing", key, err.Error())
+		}
+		c.addOrUpdateSgQueue.Forget(obj)
+		return nil
+	}(obj)
+
+	if err != nil {
+		utilruntime.HandleError(err)
+		return true
+	}
+	return true
+}
+
+func (c *Controller) processNextDeleteSgWorkItem() bool {
+	obj, shutdown := c.delSgQueue.Get()
+
+	if shutdown {
+		return false
+	}
+
+	err := func(obj interface{}) error {
+		defer c.delSgQueue.Done(obj)
+		var key string
+		var ok bool
+		if key, ok = obj.(string); !ok {
+			c.delSgQueue.Forget(obj)
+			utilruntime.HandleError(fmt.Errorf("expected string in workqueue but got %#v", obj))
+			return nil
+		}
+		if err := c.handleDeleteSg(key); err != nil {
+			c.delSgQueue.AddRateLimited(key)
+			return fmt.Errorf("error syncing '%s': %s, requeuing", key, err.Error())
+		}
+		c.delSgQueue.Forget(obj)
+		return nil
+	}(obj)
+
+	if err != nil {
+		utilruntime.HandleError(err)
+		return true
+	}
+	return true
+}
+
+func (c *Controller) initDenyAllSecurityGroup() error {
+	if err := c.ovnClient.CreateSgPortGroup(util.DenyAllSecurityGroup); err != nil {
+		return err
+	}
+	if err := c.ovnClient.CreateSgDenyAllACL(); err != nil {
+		return err
+	}
+	c.addOrUpdateSgQueue.Add(util.DenyAllSecurityGroup)
+	return nil
+}
+
+func (c *Controller) updateDenyAllSgPorts() error {
+	results, err := c.ovnClient.CustomFindEntity("logical_switch_port", []string{"_uuid", "name", "port_security"}, "external_ids:security_groups!=\"\"")
+	if err != nil {
+		klog.Errorf("failed to find logical port, %v", err)
+		return err
+	}
+	var ports []string
+	for _, ret := range results {
+		if len(ret["port_security"]) < 2 {
+			continue
+		}
+		ports = append(ports, ret["name"][0])
+	}
+	return c.ovnClient.SetPortsToPortGroup(ovs.GetSgPortGroupName(util.DenyAllSecurityGroup), ports)
+}
+
+func (c *Controller) handleAddOrUpdateSg(key string) error {
+	c.sgKeyMutex.Lock(key)
+	defer c.sgKeyMutex.Unlock(key)
+
+	// set 'deny all' for port associated with security group
+	if key == util.DenyAllSecurityGroup {
+		if err := c.updateDenyAllSgPorts(); err != nil {
+			klog.Errorf("update sg deny all policy failed. %v", err)
+			return err
+		}
+	}
+
+	sg, err := c.sgsLister.Get(key)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+
+	// check sg rules
```

Comment:
the validation check operation maybe can be abstracted to a single function

Reply:
The optimization has been completed as recommended. Thanks.

Reply:
> the validation check operation maybe can be abstracted to a single function

The optimization has been completed as recommended. Thanks.